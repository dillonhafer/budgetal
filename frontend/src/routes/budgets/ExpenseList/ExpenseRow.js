import React, { Component } from 'react';
import PropTypes from 'prop-types';

// Redux
import { connect } from 'react-redux';
import { removeExpense, selectExpense } from 'actions/budget-item-expenses';

// API
import { DeleteExpenseRequest } from '@shared/api/budget-item-expenses';

// Components
import DeleteConfirmation from 'components/DeleteConfirmation';
import {
  Text,
  Strong,
  IconButton,
  Menu,
  Popover,
  Position,
  Table,
} from 'evergreen-ui';

// Helpers
import moment from 'moment';
import { notice, error } from 'window';
import { currencyf } from '@shared/helpers';

class ExpenseRow extends Component {
  static propTypes = {
    expense: PropTypes.shape({
      id: PropTypes.number.isRequired,
      budgetItemId: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired,
      amount: PropTypes.number.isRequired,
      date: PropTypes.string.isRequired,
    }).isRequired,
    selectExpense: PropTypes.func.isRequired,
    removeExpense: PropTypes.func.isRequired,
  };

  state = {
    showDeleteConfirmation: false,
    isDeleting: false,
  };

  deleteExpense = async () => {
    const { expense } = this.props;
    this.setState({ isDeleting: true });

    try {
      const resp = await DeleteExpenseRequest(expense.id);
      if (resp && resp.ok) {
        this.deleted = true;
      }
      notice('Expense deleted');
    } catch {
      error('Something went wrong');
    } finally {
      this.setState({ isDeleting: false, showDeleteConfirmation: false });
    }
  };

  render() {
    const { expense } = this.props;
    return (
      <React.Fragment>
        <Table.Row>
          <Table.TextCell>
            {moment(expense.date, 'YYYY-MM-DD').format('MMM D, YYYY')}
          </Table.TextCell>
          <Table.TextCell flexGrow={1} flexBasis={300}>
            {expense.name}
          </Table.TextCell>
          <Table.TextCell isNumber>{currencyf(expense.amount)}</Table.TextCell>
          <Table.TextCell align="right">
            <Popover
              position={Position.BOTTOM_LEFT}
              content={
                <Menu>
                  <Menu.Group>
                    <Menu.Item
                      icon="edit"
                      onSelect={() => {
                        this.props.selectExpense(expense);
                      }}
                    >
                      Edit
                    </Menu.Item>
                  </Menu.Group>
                  <Menu.Divider />
                  <Menu.Group>
                    <Menu.Item
                      icon="trash"
                      intent="danger"
                      onSelect={() => {
                        this.setState({ showDeleteConfirmation: true });
                      }}
                    >
                      Delete
                    </Menu.Item>
                  </Menu.Group>
                </Menu>
              }
            >
              <IconButton appearance="minimal" icon="more" />
            </Popover>
          </Table.TextCell>
        </Table.Row>

        <DeleteConfirmation
          title={`Are you sure?`}
          message={
            <Text display="block">
              <Text>
                Are you sure you want to delete this expense? This cannot be
                undone.
              </Text>
              <Text display="block" marginTop={8}>
                <Strong>{expense.name}</Strong>
              </Text>
            </Text>
          }
          isShown={this.state.showDeleteConfirmation}
          isConfirmLoading={this.state.isDeleting}
          onConfirm={this.deleteExpense}
          onCloseComplete={() => {
            this.setState({ showDeleteConfirmation: false });
            if (this.deleted) {
              this.props.removeExpense(expense);
            }
          }}
        />
      </React.Fragment>
    );
  }
}

export default connect(
  null,
  dispatch => ({
    selectExpense: expense => dispatch(selectExpense(expense)),
    removeExpense: id => dispatch(removeExpense(id)),
  }),
)(ExpenseRow);
