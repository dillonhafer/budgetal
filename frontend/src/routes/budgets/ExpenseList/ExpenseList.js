import React, { Component } from 'react';
import PropTypes from 'prop-types';

// Components
import ExpenseRow from './ExpenseRow';
import { Button, Heading, Icon, Pane, Table, Text } from 'evergreen-ui';

// Helpers
import { find, groupBy, orderBy } from 'lodash';
import moment from 'moment';

class ExpenseList extends Component {
  static propTypes = {
    budget: PropTypes.shape({
      month: PropTypes.number.isRequired,
      year: PropTypes.number.isRequired,
    }),
    budgetItem: PropTypes.shape({
      id: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired,
    }).isRequired,
    budgetItemExpenses: PropTypes.array,
    selectExpense: PropTypes.func.isRequired,
  };

  newExpenseHandler = e => {
    e.preventDefault();
    const date = [
      this.props.budget.year,
      `0${this.props.budget.month}`.slice(-2),
      '01',
    ].join('-');

    this.props.selectExpense({
      name: '',
      amount: 0.0,
      date,
      budgetItemId: this.props.budgetItem.id,
    });
  };

  addExpenseLink(expenses, newFunction) {
    const disabled =
      find(expenses, expense => expense.id === null) !== undefined;
    return (
      <Button
        appearance="primary"
        onClick={newFunction}
        iconBefore="add"
        disabled={disabled}
      >
        Add an Expense
      </Button>
    );
  }

  render() {
    const budgetItemExpenses = this.props.budgetItemExpenses.filter(e => {
      return e.budgetItemId === this.props.budgetItem.id;
    });

    const sections = groupBy(
      orderBy(budgetItemExpenses, ['date', 'id'], ['desc', 'desc']),
      'date',
    );

    const sectionDates = Object.keys(sections);

    return (
      <Pane>
        <Pane
          display="flex"
          padding={16}
          background="tint2"
          marginBottom={16}
          borderRadius={3}
        >
          <Pane flex={1} alignItems="center" display="flex">
            <Heading size={600}>
              Expenses for {this.props.budgetItem.name}
            </Heading>
          </Pane>
          <Pane>
            {this.addExpenseLink(budgetItemExpenses, this.newExpenseHandler)}
          </Pane>
        </Pane>

        <Pane
          display={sectionDates.length > 0 ? 'none' : 'flex'}
          padding={16}
          border="muted"
          marginBottom={16}
          borderRadius={3}
        >
          <Pane
            flex={1}
            alignItems="center"
            justifyContent="center"
            display="flex"
          >
            <Icon icon="dollar" marginRight={4} />
            <Text>There are no expenses yet</Text>
          </Pane>
        </Pane>

        <Table
          background="white"
          display={sectionDates.length > 0 ? '' : 'none'}
          border="muted"
          borderRadius="3"
        >
          <Table.Head accountForScrollbar>
            <Table.TextHeaderCell>Date</Table.TextHeaderCell>
            <Table.TextHeaderCell flexGrow={1} flexBasis={300}>
              Name
            </Table.TextHeaderCell>
            <Table.TextHeaderCell>Amount</Table.TextHeaderCell>
            <Table.TextHeaderCell />
          </Table.Head>
          <Table.VirtualBody height={500}>
            {sectionDates.map(section => {
              return (
                <React.Fragment key={section}>
                  <Table.Head accountForScrollbar>
                    <Table.TextHeaderCell flexBasis={300} flexGrow={1}>
                      {moment(section, 'YYYY-MM-DD').format(
                        'dddd - MMM DD, YYYY',
                      )}
                    </Table.TextHeaderCell>
                    <Table.TextHeaderCell />
                    <Table.TextHeaderCell />
                    <Table.TextHeaderCell />
                  </Table.Head>
                  <Table.Body>
                    {sections[section].map(expense => (
                      <ExpenseRow key={expense.id} expense={expense} />
                    ))}
                  </Table.Body>
                </React.Fragment>
              );
            })}
          </Table.VirtualBody>
        </Table>
      </Pane>
    );
  }
}

export default ExpenseList;
