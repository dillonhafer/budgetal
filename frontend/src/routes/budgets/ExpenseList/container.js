// Redux
import { connect } from 'react-redux';
import { selectExpense } from 'actions/budget-item-expenses';
import ExpenseList from './ExpenseList';

export default connect(
  state => ({
    ...state.budget,
  }),
  dispatch => ({
    selectExpense: expense => dispatch(selectExpense(expense)),
  }),
)(ExpenseList);
