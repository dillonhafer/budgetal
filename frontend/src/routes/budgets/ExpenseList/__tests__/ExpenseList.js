import React from 'react';
import { shallow } from 'enzyme';
import ExpenseList from '../ExpenseList';

const defaultProps = {
  budget: {
    year: '1996',
    month: '1',
  },
  budgetItem: {
    id: 1,
    name: 'Item Name',
  },
  budgetItemExpenses: [
    {
      id: 1,
      budgetItemId: 1,
      amount: 3000.0,
      name: '',
      date: '2019-01-01',
    },
  ],
  selectExpense: jest.fn(),
};

describe('ExpenseList', () => {
  it('renders a ExpenseList', () => {
    const comp = shallow(<ExpenseList {...defaultProps} />);
    expect(comp).toMatchSnapshot();
  });
});
