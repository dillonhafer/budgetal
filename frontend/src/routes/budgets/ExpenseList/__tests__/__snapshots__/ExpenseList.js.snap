// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpenseList renders a ExpenseList 1`] = `
<withTheme(Pane)>
  <withTheme(Pane)
    background="tint2"
    borderRadius={3}
    display="flex"
    marginBottom={16}
    padding={16}
  >
    <withTheme(Pane)
      alignItems="center"
      display="flex"
      flex={1}
    >
      <withTheme(Heading)
        size={600}
      >
        Expenses for 
        Item Name
      </withTheme(Heading)>
    </withTheme(Pane)>
    <withTheme(Pane)>
      <withTheme(Button)
        appearance="primary"
        disabled={false}
        iconBefore="add"
        onClick={[Function]}
      >
        Add an Expense
      </withTheme(Button)>
    </withTheme(Pane)>
  </withTheme(Pane)>
  <withTheme(Pane)
    border="muted"
    borderRadius={3}
    display="none"
    marginBottom={16}
    padding={16}
  >
    <withTheme(Pane)
      alignItems="center"
      display="flex"
      flex={1}
      justifyContent="center"
    >
      <withTheme(Icon)
        icon="dollar"
        marginRight={4}
      />
      <withTheme(Text)>
        There are no expenses yet
      </withTheme(Text)>
    </withTheme(Pane)>
  </withTheme(Pane)>
  <Table
    background="white"
    border="muted"
    borderRadius="3"
    display=""
  >
    <TableHead
      accountForScrollbar={true}
      height={32}
    >
      <TextTableHeaderCell>
        Date
      </TextTableHeaderCell>
      <TextTableHeaderCell
        flexBasis={300}
        flexGrow={1}
      >
        Name
      </TextTableHeaderCell>
      <TextTableHeaderCell>
        Amount
      </TextTableHeaderCell>
      <TextTableHeaderCell />
    </TableHead>
    <TableVirtualBody
      allowAutoHeight={false}
      defaultHeight={48}
      height={500}
      overscanCount={5}
      useAverageAutoHeightEstimation={true}
    >
      <TableHead
        accountForScrollbar={true}
        height={32}
      >
        <TextTableHeaderCell
          flexBasis={300}
          flexGrow={1}
        >
          Tuesday - Jan 01, 2019
        </TextTableHeaderCell>
        <TextTableHeaderCell />
        <TextTableHeaderCell />
        <TextTableHeaderCell />
      </TableHead>
      <TableBody>
        <Connect(ExpenseRow)
          expense={
            Object {
              "amount": 3000,
              "budgetItemId": 1,
              "date": "2019-01-01",
              "id": 1,
              "name": "",
            }
          }
          key="1"
        />
      </TableBody>
    </TableVirtualBody>
  </Table>
</withTheme(Pane)>
`;
